# Collectible Car Inventory Management

Welcome to a detailed exploration of a unique dataset featuring collectible cars from my father's cherished collection. This notebook chronicles our efforts to transform a cluttered assortment of data into a streamlined, structured inventory. Such a transformation is essential, forming the foundation for a sophisticated inventory management system designed to enhance both the efficiency and effectiveness of how these collectible gems are handled.

# Project Outline

1. **Data Exploration:** Dive into the dataset to grasp its basic layout and constituent elements.
2. **Data Restructuring:** Tailor the dataset to fit the practical needs of an inventory management system, enhancing usability.
3. **Data Cleaning:** Tackle any data inconsistencies, fill missing values, and correct inaccuracies to polish the dataset.

Our ultimate objective is to refine the dataset to such an extent that it not only offers a transparent snapshot of the collection but also becomes fully integrated with an inventory management application, built with Google Appsheet. This integration will facilitate straightforward updates and maintenance, bolstering both the enjoyment of the collection in the present and its legacy in the future.

Embark on this journey with me as we turn a raw dataset into an indispensable tool for managing a collection of automotive treasures!


# 1. Data Exploration

## 1.1 Initial Data Loading and Inspection

We'll begin our journey by loading the dataset from an Excel file—kindly provided by my father, who in this case, plays the role of both client and collector. Our first step is to gain a preliminary understanding of the dataset's structure and content. This initial peek will help us identify the type of data we're dealing with and set the stage for the detailed analysis that follows.

In [1]:
import pandas as pd

# Filepath
file_path = '/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx'

# Load the data from the Excel file
data = pd.read_excel(file_path)

# Display the structure of the DataFrame
print("Data Info:")
data.info()

# Display the first few rows to understand the data better
print("First few rows of the dataset:")
display(data.head())


Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 410 entries, 0 to 409
Columns: 181 entries, Altaya to Road Champions
dtypes: object(181)
memory usage: 579.9+ KB
First few rows of the dataset:


Altaya  \
0            Matra Simca Rancho 1978   
1  Citroen DS 19 Amsterdam Taxi 1958   
2          Tuk tuk Bangkok taxi 1980   
3                  Simca 1301 S 1972   
4   Peugeot 203 Casablanca Taxi 1960   

                             Opel Collection journal  \
0               Opel Diplomat V8 Limousine 1964-1967   
1  Opel Rekord D 2.1 liter 1973-1977 - white/blac...   
2                         Opel Lotus Omega 1989-1992   
3                         Opel Kapitan '38 1938-1940   
4                  Opel 10/40 PS Modell 80 1925-1929   

                    Opel Collection Designer-Serie  \
0          Opel GT '1968 - silver - Erhard Schnell   
1  Opel Record A Cabrio - Clare MacKichan - silver   
2                                              NaN   
3                                              NaN   
4                                              NaN   

            Deutsche Liebhaber-autos  \
0  Borgward Isabella Coupé 1957-1961   
1  Volkswagen 1302 LS Limousine 1972   
2                                NaN   
3                                NaN   
4                                NaN   

  Auto Plus - Les classiques de l'automobile  \
0                      Panhard Dyna Z - 1958   
1                         Simca Marly - 1959   
2                 Talbot 150 GT - 1982 - red   
3    Renault Dauphine Ondine - 1961 - silver   
4                   Citroen LNA - 1985 - red   

                  Mercedes-Benz journal series (Ixo)  \
0  Mercedes-Benz 600 Pullman LWB (1963) - Midnigh...   
1            Mercedes-Benz 500 K Autobahnkurier 1934   
2                 Mercedes-Benz 300 SC 1956 Roadster   
3                 Mercedes-Benz 280 SL 1963 Roadster   
4  Mercedes-Benz 220 SE W111 Heckflosse 1959-1965...   

                                                 Ixo  \
0  Range Rover Vogue 1980 Paris-Dakar (Dakar 1979...   
1                    Lagonda LG6 Drophead Coupe 1938   
2     Renault Furgonette 300KG Juvaquatre - Michelin   
3                              Maserati Khamsin 1972   
4        Ford Mk IV #4 Le Mans 1967 (D.Hulme-L.Ruby)   

                                        Premium X  \
0                  Buick Riviera Coupè 1972 green   
1  Stutz Blackhawk Convertible 1971 with hard top   
2                          1973 Chevrolet Bel Air   
3    1956 Volvo Amazon 120 sedam - red/white roof   
4                           1972 Ford Gran Torino   

                            Cofradis Limited Edition  \
0  Bugatti 35B Gagnante GP de Monaco 1929 #12 Pil...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                       British Champions' Cars (Ixo)  ...  \
0  Mercedes-Benz 300 SLR - Stirling Moss/Denis Je...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

                     Beer Trucks  \
0  Robur Garant Lübzer Pils 1:64   
1   Framo Wernesgrüner Pils 1:64   
2                            NaN   
3                            NaN   
4                            NaN   

                                               Luppa  \
0  SMG Buggy Red Bull #303 Carlos Sainz/Timo Gott...   
1  Peugeot 2008 DKR15+ #328 R.Dumas/F.Borsotto - ...   
2  Toyota Hilux V8 #330 Alejandro Yacopini/Daniel...   
3  Toyota Hilux V8 #305 Nani Roma/Alex Haro Bravo...   
4  Renault Duster #317 Emiliano Spataro/Benjamin ...   

                                  Direkt Collections  \
0  Bernard 110MB - Centrale Electrique Cirque Pinder   
1            Unic Cuisine Cirque Pinder - red/yellow   
2                                                NaN   
3                                                NaN   
4                                                NaN   

         

## 1.1.1 Brief overview based on the first 5 rows:

- The dataset from the "Cars catalogue Main_RAW.xlsx" file comprises **410 entries** across **181 columns**.
- Each column represents a distinct collection or series of car models.
- The dataset showcases an eclectic mix of car collections, including "Altaya," "Opel Collection journal," "Opel Collection Designer-Serie," "Deutsche Liebhaber-autos," "Auto Plus - Les classiques de l'automobile," "Mercedes-Benz journal series (Ixo)," among others.
- Entries detail specific car models, occasionally accompanied by particulars such as model year, color, or specific edition information, albeit without uniform formatting. Examples include "Matra Simca Rancho 1978", "Opel 10/40 PS Modell 80 1925-1929", "Mercedes-Benz 280 SL 1963 Roadster", etc.

## 1.1.2 Observations from the Data

- **Sparsity:** A significant number of columns exhibit missing values, suggesting incomplete records.
- **Inconsistency:** There is considerable variation in descriptions, blending information about the model, year, and color in differing formats.

## Next Steps:

- **Data Structuring:** The plan involves reformatting the data into a more analysis-conducive layout, potentially merging similar columns or restructuring the dataset based on specific characteristics like model year, make, or collection.
- **Data Cleaning:** This step will address the missing data, standardize textual entries, and possibly reformat the data to enhance usability.
- **Analysis/Visualization Preparation:** We will identify crucial variables for detailed analysis and prepare the dataset for integration into visualization tools like Looker Studio.

# 2. Data Restructuring

We are now set to restructure the dataset into a more streamlined format. The new structure will feature just two columns: one for the manufacturer's name and the other for the full item name. This simplification will make the dataset not only easier to handle but also more amenable for deeper analysis and visualization.

In this process, we will create a new DataFrame where each row represents a unique collectible, clearly identified by its manufacturer and full name. This step is designed to facilitate the extraction of key details such as brand, model, year, and color, enhancing the dataset's utility for detailed analysis and efficient data management.


In [2]:
import pandas as pd

# Filepath
file_path = '/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx'

# Load the dataset
data = pd.read_excel(file_path)

# Melt the DataFrame to reformat it into two columns: 'Model_Manufacturer_name' and 'Collectible_item_full_name'
melted_data = data.melt(var_name='Model_Manufacturer_name', value_name='Collectible_item_full_name')

# Remove rows where 'Collectible_item_full_name' is null as they do not provide useful information
cleaned_data = melted_data.dropna(subset=['Collectible_item_full_name'])

# Show the first few rows of the newly structured DataFrame and summary information
print("Summary Information of Cleaned Data:")
cleaned_data_info = cleaned_data.info()

print("\nFirst Few Rows of Cleaned Data:")
cleaned_data_head = cleaned_data.head()

# If you need to save or export the cleaned data:
cleaned_data.to_csv('/kaggle/working/cleaned_data_v1_2columns.csv', index=False)

# Display the cleaned data head
display(cleaned_data_head)


Summary Information of Cleaned Data:
<class 'pandas.core.frame.DataFrame'>
Index: 4045 entries, 0 to 73801
Data columns (total 2 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Model_Manufacturer_name     4045 non-null   object
 1   Collectible_item_full_name  4045 non-null   object
dtypes: object(2)
memory usage: 94.8+ KB

First Few Rows of Cleaned Data:


Model_Manufacturer_name         Collectible_item_full_name
0                  Altaya            Matra Simca Rancho 1978
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958
2                  Altaya          Tuk tuk Bangkok taxi 1980
3                  Altaya                  Simca 1301 S 1972
4                  Altaya   Peugeot 203 Casablanca Taxi 1960

The transformation is complete! Our dataset now neatly organizes **4,045 entries** into two distinct columns: **'Model_Manufacturer_name'** and **'Collectible_item_full_name'**. This format sharpens the focus, with each entry distinctly representing a collectible item. The 'Model_Manufacturer_name' column tags each item with its manufacturer or series, while the 'Collectible_item_full_name' provides enriched details, encompassing elements like the brand name, model, and occasionally the year and color. This restructured dataset lays a solid foundation for extracting precise attributes and diving deeper into data analysis.

## 2.1 Extracting Relevant Information

To further refine our dataset, we will now focus on extracting key details such as the brand, model, and, where possible, the year and color of each collectible item. This step is crucial as it enhances the dataset's utility for detailed analysis and reporting. The structured data will be outlined as follows:

* **Model_Manufacturer_name:** Identifies the manufacturer or series.
* **Collectible_item_full_name:** Lists the complete name of the collectible item as it appears in the records.
* **Brand:** The brand name derived from the full item description.
* **Model:** Detailed model information extracted from the item description.
* **Year:** The year or range of years associated with the collectible item, when available.
* **Color:** Color information, extracted if specified within the item description.


In [3]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Function to extract brand, model, and year from the collectible item full name
def extract_info(text):
    pattern = r'^(?P<Brand>[\w\s-]+?)\s(?P<Model>.*?)\s(?P<Year>\d{4}(?:-\d{4})?)'
    match = re.match(pattern, str(text))
    if match:
        return match.group('Brand').strip(), match.group('Model').strip(), match.group('Year')
    else:
        return None, None, None

# Initialize an empty list to store the reshaped data
reshaped_rows = []

# Iterate through each column of the original DataFrame
for col in data.columns:
    # Extract manufacturer name from the column name
    manufacturer_name = col.split('_')[0]
    # Iterate through each item in the column
    for item in data[col]:
        # Check if the item is not empty
        if pd.notna(item):
            brand, model, year = extract_info(item)
            # Append a new row to the reshaped data list
            reshaped_rows.append({'Model_Manufacturer_name': manufacturer_name,
                                  'Collectible_item_full_name': item,
                                  'Brand': brand, 'Model': model, 'Year': year})

# Convert the list of dictionaries to a DataFrame
reshaped_data = pd.DataFrame(reshaped_rows)

# Filter for valid entries where 'Model_Manufacturer_name' is not missing
valid_data = reshaped_data[reshaped_data['Model_Manufacturer_name'].notna()]

# Count missing 'Brand' values in the valid data
missing_brands_count = valid_data[valid_data['Brand'].isnull()].shape[0]
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brands_count}")

# Count missing 'Year' values in the valid data
missing_years_count = valid_data[valid_data['Year'].isnull()].shape[0]
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_years_count}")

# If you need to save or export the cleaned data:
cleaned_data.to_csv('/kaggle/working/cleaned_data_v2_multiCol_1675-missing-brand.csv', index=False)

# Display the reshaped data
display(reshaped_data)

Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 1675
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 1675


Model_Manufacturer_name  \
0                     Altaya   
1                     Altaya   
2                     Altaya   
3                     Altaya   
4                     Altaya   
...                      ...   
4040                Motorama   
4041                Motorama   
4042                   Vroom   
4043           Kager Edition   
4044          Road Champions   

                             Collectible_item_full_name          Brand  \
0                               Matra Simca Rancho 1978          Matra   
1                     Citroen DS 19 Amsterdam Taxi 1958        Citroen   
2                             Tuk tuk Bangkok taxi 1980            Tuk   
3                                     Simca 1301 S 1972          Simca   
4                      Peugeot 203 Casablanca Taxi 1960        Peugeot   
...                                                 ...            ...   
4040  Dallara SP2 Oreca #15 O.Beretta/P.Lamy/E.Comas...        Dallara   
4041          BMW Z4 M Coupè Racing #47 Heinz Schmersal           None   
4042  Porsche 910 8 cyl Spyder #154, Nurburgring 196...        Porsche   
4043  Mercedes-Benz 300 SLR #19 Fangio/Moss, Le Mans...  Mercedes-Benz   
4044        1969 Dodge Super Bee - met. Grey/black roof           None   

                                                  Model  Year  
0                                          Simca Rancho  1978  
1                                  DS 19 Amsterdam Taxi  1958  
2                                      tuk Bangkok taxi  1980  
3                                                1301 S  1972  
4                                   203 Casablanca Taxi  1960  
...                                                 ...   ...  
4040  SP2 Oreca #15 O.Beretta/P.Lamy/E.Comas, 5th Le...  2002  
4041                                               None  None  
4042                 910 8 cyl Spyder #154, Nurburgring  1967  
4043                   300 SLR #19 Fangio/Moss, Le Mans  1955  
4044                                               None  None  

[4045 rows x 5 columns]

## 2.2 Filtering Missing Values

Upon closer inspection, we encounter a significant challenge: 1675 entries still lack complete information in the 'Brand', 'Model', and 'Year' columns. This issue stems from the varied formats of the item names, which our initial regular expression failed to uniformly interpret. Instead of removing these entries, we'll take a different approach:

1. **Identify and Isolate Incomplete Entries:** We will filter and isolate entries missing essential data, which will help in pinpointing the inconsistencies.
2. **Analyze Naming Variations:** By analyzing these variations, we can adjust our data extraction techniques to capture more diverse naming conventions.

### Filtered Data:

We will provide a list of these filtered entries to further analyze and refine our regular expression, ensuring a more comprehensive data capture in future iterations.


In [4]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Function to extract brand, model, and year from the collectible item full name
def extract_info(text):
    pattern = r'^(?P<Brand>[\w\s-]+?)\s(?P<Model>.*?)\s(?P<Year>\d{4}(?:-\d{4})?)'
    match = re.match(pattern, str(text))
    if match:
        return match.group('Brand').strip(), match.group('Model').strip(), match.group('Year')
    else:
        return None, None, None

# Initialize an empty list to store the reshaped data
reshaped_rows = []

# Iterate through each column of the original DataFrame
for col in data.columns:
    # Extract manufacturer name from the column name
    manufacturer_name = col.split('_')[0]
    # Iterate through each item in the column
    for item in data[col]:
        # Check if the item is not empty
        if pd.notna(item):
            brand, model, year = extract_info(item)
            # Append a new row to the reshaped data list
            reshaped_rows.append({'Model_Manufacturer_name': manufacturer_name,
                                  'Collectible_item_full_name': item,
                                  'Brand': brand, 'Model': model, 'Year': year})

# Convert the list of dictionaries to a DataFrame
reshaped_data = pd.DataFrame(reshaped_rows)

# Filter for valid entries where 'Model_Manufacturer_name' is not missing
valid_data = reshaped_data[reshaped_data['Model_Manufacturer_name'].notna()]

# Count missing 'Brand' values in the valid data
missing_brands_count = valid_data[valid_data['Brand'].isnull()].shape[0]
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brands_count}")

# Count missing 'Year' values in the valid data
missing_years_count = valid_data[valid_data['Year'].isnull()].shape[0]
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_years_count}")

# Filter entries without a brand
entries_without_brand = valid_data[valid_data['Brand'].isnull()]

# Display entries without a brand
print("Entries without a brand:")
display(entries_without_brand)

# If you need to save or export the cleaned data:
cleaned_data.to_csv('/kaggle/working/cleaned_data_v2_list_missing-brand-only.csv', index=False)



Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 1675
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 1675
Entries without a brand:


Model_Manufacturer_name                   Collectible_item_full_name  \
7                     Altaya                    Renault Furgonette 300 kg   
8                     Altaya         Peugeot J5 Furgone 'Touring Secours'   
18                    Altaya                               Mercedes 540 K   
20                    Altaya                               Maserati 250 F   
21                    Altaya                   Pagani Zonda C12S - silver   
...                      ...                                          ...   
4036      Direkt Collections      Unic Cuisine Cirque Pinder - red/yellow   
4038                 Camater     Rolls Royce Silver Spirit II - gold met.   
4039               Centauria   Piaggio Vespa TAP (1956) - military - 1:18   
4041                Motorama    BMW Z4 M Coupè Racing #47 Heinz Schmersal   
4044          Road Champions  1969 Dodge Super Bee - met. Grey/black roof   

     Brand Model  Year  
7     None  None  None  
8     None  None  None  
18    None  None  None  
20    None  None  None  
21    None  None  None  
...    ...   ...   ...  
4036  None  None  None  
4038  None  None  None  
4039  None  None  None  
4041  None  None  None  
4044  None  None  None  

[1675 rows x 5 columns]

With all entries lacking 'Brand', 'Model', and 'Year' neatly catalogued, we face a daunting task. Manually reviewing the 1675 records would be both time-consuming and inefficient. To streamline this process, we've opted for a more technologically savvy approach:

### Leveraging AI for Data Enhancement:

We will export this list of incomplete entries to a file specifically for analysis by ChatGPT4. This advanced AI model will swiftly analyze the naming variations and suggest enhancements to our regular expression. This step aims to minimize the missing values significantly, optimizing our dataset for further analysis and use.

## 2.2 Reducing the Missing Values

### Enhancing Data Integrity:

Following a thorough analysis, we've incorporated new variations into our regular expression patterns. This enhancement is designed to capture more detailed attributes from each entry, thereby reducing the number of missing values.

In [5]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Function to extract brand, model, and year, with enhanced regex
def extract_info_improved(text):
    # Normalize the text by removing known noise patterns and handling edge cases
    text = str(text)
    text = re.sub(r'\s+\(.*?\)', '', text)  # Remove any content inside parentheses
    text = re.sub(r'\s-\s.*', '', text)     # Remove descriptions after a dash
    text = re.sub(r"\b(?<!\d)(?!\d{4})\d+\b", "", text)  # Remove isolated numbers that are not part of a four-digit year
    text = text.replace(',', '')  # Remove commas that might be used as separators

    # Enhanced regex pattern to handle various cases
    pattern = (
        r'^(?P<Brand>\D+?)'             # Capture the brand as non-digit characters at the start
        r'\s+(?P<Model>.*?)'            # Capture the model which might include numbers
        r'(\s+(?P<Year>\d{4}))?$'       # Optionally capture a four-digit year at the end
    )
    match = re.match(pattern, text)
    if match:
        brand = match.group('Brand').strip()
        model = match.group('Model').strip()
        year = match.group('Year') if match.group('Year') else None
        return brand, model, year
    return None, None, None

# Process each item, reshape data
reshaped_rows = []
for col in data.columns:
    manufacturer_name = col.split('_')[0]
    for item in data[col]:
        if pd.notna(item):
            brand, model, year = extract_info_improved(item)
            reshaped_rows.append({
                'Model_Manufacturer_name': manufacturer_name,
                'Collectible_item_full_name': item,
                'Brand': brand, 'Model': model, 'Year': year
            })

# Convert reshaped_rows into a DataFrame
reshaped_data = pd.DataFrame(reshaped_rows)

# Filter for valid entries where 'Model_Manufacturer_name' is not empty
valid_data = reshaped_data[reshaped_data['Model_Manufacturer_name'].notna()]

# Count and print missing 'Brand' and 'Year' values excluding missing 'Model_Manufacturer_name'
missing_brands_count = valid_data[valid_data['Brand'].isnull()].shape[0]
missing_years_count = valid_data[valid_data['Year'].isnull()].shape[0]
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brands_count}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_years_count}")

# Convert to DataFrame, export, and display
reshaped_data.to_csv('/kaggle/working/cleaned_data_v3_multiCol_292-missing-brand.csv', index=False)

display(reshaped_data.head())


Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 292
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 2954


Model_Manufacturer_name         Collectible_item_full_name    Brand  \
0                  Altaya            Matra Simca Rancho 1978    Matra   
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958  Citroen   
2                  Altaya          Tuk tuk Bangkok taxi 1980      Tuk   
3                  Altaya                  Simca 1301 S 1972    Simca   
4                  Altaya   Peugeot 203 Casablanca Taxi 1960  Peugeot   

                Model  Year  
0        Simca Rancho  1978  
1  DS  Amsterdam Taxi  1958  
2    tuk Bangkok taxi  1980  
3              1301 S  1972  
4     Casablanca Taxi  1960

We've made significant progress in addressing incomplete data. The count of entries missing the 'Brand' detail has been reduced dramatically to just 292. This improvement provides a more solid dataset to work with moving forward.

In [6]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Common color names for regex (this list can be extended)
colors = "black|white|red|green|blue|yellow|silver|grey|orange|purple|gold|bronze|brown"

# Function to extract brand, model, year, and color, with enhanced regex
def extract_info_improved(text):
    # Normalize the text by removing known noise patterns and handling edge cases
    text = str(text)
    text = re.sub(r'\s+\(.*?\)', '', text)  # Remove any content inside parentheses
    text = re.sub(r'\s-\s.*', '', text)     # Remove descriptions after a dash
    text = re.sub(r"\b(?<!\d)(?!\d{4})\d+\b", "", text)  # Remove isolated numbers that are not part of a four-digit year
    text = text.replace(',', '')  # Remove commas that might be used as separators

    # Enhanced regex pattern to handle various cases including color
    pattern = (
        rf'^(?P<Brand>\D+?)'             # Capture the brand as non-digit characters at the start
        r'\s+(?P<Model>.*?)'             # Capture the model which might include numbers
        r'(\s+(?P<Year>\d{{4}}))?'       # Optionally capture a four-digit year at the end
        rf'(\s+(?P<Color>{colors}))?$'   # Optionally capture a color at the end
    )
    match = re.match(pattern, text)
    if match:
        brand = match.group('Brand').strip()
        model = match.group('Model').strip()
        year = match.group('Year') if match.group('Year') else None
        color = match.group('Color') if 'Color' in match.groupdict() and match.group('Color') else None
        return brand, model, year, color
    return None, None, None, None

# Process each item, reshape data
reshaped_rows = []
for col in data.columns:
    manufacturer_name = col.split('_')[0]
    for item in data[col]:
        if pd.notna(item):
            brand, model, year, color = extract_info_improved(item)
            reshaped_rows.append({
                'Model_Manufacturer_name': manufacturer_name,
                'Collectible_item_full_name': item,
                'Brand': brand,
                'Model': model,
                'Year': year,
                'Color': color
            })

# Convert reshaped_rows into a DataFrame
reshaped_data = pd.DataFrame(reshaped_rows)

# Filter for valid entries where 'Model_Manufacturer_name' is not empty
valid_data = reshaped_data[reshaped_data['Model_Manufacturer_name'].notna()]

# Count and print missing 'Brand' and 'Year' values excluding missing 'Model_Manufacturer_name'
missing_brands_count = valid_data[valid_data['Brand'].isnull()].shape[0]
missing_years_count = valid_data[valid_data['Year'].isnull()].shape[0]
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brands_count}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_years_count}")

# Display the head of the DataFrame to confirm the extraction
display(reshaped_data.head())


Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 292
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 4045


Model_Manufacturer_name         Collectible_item_full_name    Brand  \
0                  Altaya            Matra Simca Rancho 1978    Matra   
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958  Citroen   
2                  Altaya          Tuk tuk Bangkok taxi 1980      Tuk   
3                  Altaya                  Simca 1301 S 1972    Simca   
4                  Altaya   Peugeot 203 Casablanca Taxi 1960  Peugeot   

                     Model  Year Color  
0        Simca Rancho 1978  None  None  
1  DS  Amsterdam Taxi 1958  None  None  
2    tuk Bangkok taxi 1980  None  None  
3              1301 S 1972  None  None  
4     Casablanca Taxi 1960  None  None

In [7]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Common color names for regex (this list can be extended)
colors = "black|white|red|green|blue|yellow|silver|grey|orange|purple|gold|bronze|brown"

# Function to extract brand, model, year, and color, with enhanced regex
def extract_info_improved(text):
    # Normalize the text by removing known noise patterns and handling edge cases
    text = str(text)
    text = re.sub(r'\s+\(.*?\)', '', text)  # Remove any content inside parentheses
    text = re.sub(r'\s-\s.*', '', text)     # Remove descriptions after a dash
    text = re.sub(r"\b(?<!\d)(?!\d{4})\d+\b", "", text)  # Remove isolated numbers that are not part of a four-digit year
    text = text.replace(',', '')  # Remove commas that might be used as separators

    # Enhanced regex pattern to handle various cases including color
    pattern = (
        rf'^(?P<Brand>\D+?)'             # Capture the brand as non-digit characters at the start
        r'\s+(?P<Model>.*?)'             # Capture the model which might include numbers
        r'(\s+(?P<Year>\d{{4}}))?'       # Optionally capture a four-digit year at the end
        rf'(\s+(?P<Color>{colors}))?$'   # Optionally capture a color at the end
    )
    match = re.match(pattern, text)
    if match:
        brand = match.group('Brand').strip()
        model = match.group('Model').strip()
        year = match.group('Year') if match.group('Year') else None
        color = match.group('Color') if 'Color' in match.groupdict() and match.group('Color') else None
        return brand, model, year, color
    return None, None, None, None

# Process each item, reshape data
reshaped_rows = []
for col in data.columns:
    manufacturer_name = col.split('_')[0]
    for item in data[col]:
        if pd.notna(item):
            brand, model, year, color = extract_info_improved(item)
            reshaped_rows.append({
                'Model_Manufacturer_name': manufacturer_name,
                'Collectible_item_full_name': item,
                'Brand': brand,
                'Model': model,
                'Year': year,
                'Color': color
            })

# Convert reshaped_rows into a DataFrame
reshaped_data = pd.DataFrame(reshaped_rows)

# Filter for valid entries where 'Model_Manufacturer_name' is not empty
valid_data = reshaped_data[reshaped_data['Model_Manufacturer_name'].notna()]

# Count and print missing 'Brand', 'Year', and 'Color' values excluding missing 'Model_Manufacturer_name'
missing_brands_count = valid_data[valid_data['Brand'].isnull()].shape[0]
missing_years_count = valid_data[valid_data['Year'].isnull()].shape[0]
missing_colors_count = valid_data[valid_data['Color'].isnull()].shape[0]
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brands_count}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_years_count}")
print(f"Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': {missing_colors_count}")

# Display the head of the DataFrame to confirm the extraction
display(reshaped_data.head())


Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 292
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 4045
Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': 3900


Model_Manufacturer_name         Collectible_item_full_name    Brand  \
0                  Altaya            Matra Simca Rancho 1978    Matra   
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958  Citroen   
2                  Altaya          Tuk tuk Bangkok taxi 1980      Tuk   
3                  Altaya                  Simca 1301 S 1972    Simca   
4                  Altaya   Peugeot 203 Casablanca Taxi 1960  Peugeot   

                     Model  Year Color  
0        Simca Rancho 1978  None  None  
1  DS  Amsterdam Taxi 1958  None  None  
2    tuk Bangkok taxi 1980  None  None  
3              1301 S 1972  None  None  
4     Casablanca Taxi 1960  None  None

## 2.3 Adding IDs with More Year Information

In this phase, our goal is to enrich the dataset by focusing on extracting year information more reliably. Using regular expressions, we'll parse out year data from the entries wherever available. To streamline further operations and ensure each record can be uniquely identified, we will also introduce a unique ID for each item. This step is crucial for simplifying the integration of data across the following processing stages.

In [8]:
import pandas as pd
import re
from IPython.display import display

# Simulating the loading of the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Function to extract brand, model, and year with robust error handling
def extract_info_improved(text):
    text = str(text)  # Ensure text is a string
    pattern = (
        r'^(?P<Brand>\D+?)'             
        r'\s+(?P<Model>.*?)'            
        r'(\s+(?P<Year>\d{4}))?$'      
    )
    match = re.match(pattern, text)
    if match:
        brand = match.group('Brand').strip() if match.group('Brand') else None
        model = match.group('Model').strip() if match.group('Model') else None
        year = match.group('Year') if match.group('Year') else None
        return brand, model, year
    else:
        return None, None, None

# Initialize counter for unique ID generation
counter = 1

# Process each item, reshape data
reshaped_rows = []
for col in data.columns:
    manufacturer_name = col.split('_')[0]
    for item in data[col]:
        if pd.notna(item):
            brand, model, year = extract_info_improved(item)
            reshaped_rows.append({
                'Model_Manufacturer_name': manufacturer_name,
                'Collectible_item_full_name': item,
                'Brand': brand, 'Model': model, 'Year': year,
                'id': f"{manufacturer_name}-{item}-{counter:04d}"
            })
            counter += 1

# Convert to DataFrame
df_a = pd.DataFrame(reshaped_rows)

# Reporting missing data and total entries
total_entries = len(df_a)
missing_brand = df_a['Brand'].isnull().sum()
missing_year = df_a['Year'].isnull().sum()

print(f"Total number of entries: {total_entries}")
print(f"Number of entries with missing 'Brand': {missing_brand}")
print(f"Number of entries with missing 'Year': {missing_year}")

# Display the DataFrame to verify results
display(df_a.head())


Total number of entries: 4045
Number of entries with missing 'Brand': 310
Number of entries with missing 'Year': 3550


Model_Manufacturer_name         Collectible_item_full_name    Brand  \
0                  Altaya            Matra Simca Rancho 1978    Matra   
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958  Citroen   
2                  Altaya          Tuk tuk Bangkok taxi 1980      Tuk   
3                  Altaya                  Simca 1301 S 1972    Simca   
4                  Altaya   Peugeot 203 Casablanca Taxi 1960  Peugeot   

                  Model  Year                                             id  
0          Simca Rancho  1978            Altaya-Matra Simca Rancho 1978-0001  
1  DS 19 Amsterdam Taxi  1958  Altaya-Citroen DS 19 Amsterdam Taxi 1958-0002  
2      tuk Bangkok taxi  1980          Altaya-Tuk tuk Bangkok taxi 1980-0003  
3                1301 S  1972                  Altaya-Simca 1301 S 1972-0004  
4   203 Casablanca Taxi  1960   Altaya-Peugeot 203 Casablanca Taxi 1960-0005

## 2.4 Adding IDs with More Color Information

Here, we extend our data extraction efforts to include color details, enhancing the dataset's descriptive power. By refining our regex patterns, we effectively capture color names embedded within the entries. This addition not only enriches the dataset with visual descriptors of the collectible cars but also boosts the overall appeal and functionality of the catalogue.

In [9]:
import pandas as pd
import re
from IPython.display import display

# Read the dataset
data = pd.read_excel('/kaggle/input/cars-catalogue-main-raw/Cars catalogue Main_RAW.xlsx')

# Common color names for regex (this list can be extended)
colors = "black|white|red|green|blue|yellow|silver|grey|orange|purple|gold|bronze|brown"

# Function to extract brand, model, year, and color, with enhanced regex
def extract_info_improved(text):
    # Normalize the text by removing known noise patterns and handling edge cases
    text = str(text)
    text = re.sub(r'\s+\(.*?\)', '', text)  # Remove any content inside parentheses
    text = re.sub(r'\s-\s.*', '', text)     # Remove descriptions after a dash
    text = re.sub(r"\b(?<!\d)(?!\d{4})\d+\b", "", text)  # Remove isolated numbers that are not part of a four-digit year
    text = text.replace(',', '')  # Remove commas that might be used as separators

    # Enhanced regex pattern to handle various cases including color
    pattern = (
        rf'^(?P<Brand>\D+?)'             # Capture the brand as non-digit characters at the start
        r'\s+(?P<Model>.*?)'             # Capture the model which might include numbers
        r'(\s+(?P<Year>\d{{4}}))?'       # Optionally capture a four-digit year at the end
        rf'(\s+(?P<Color>{colors}))?$'   # Optionally capture a color at the end
    )
    match = re.match(pattern, text)
    if match:
        brand = match.group('Brand').strip()
        model = match.group('Model').strip()
        year = match.group('Year') if match.group('Year') else None
        color = match.group('Color') if 'Color' in match.groupdict() and match.group('Color') else None
        return brand, model, year, color
    else:
        return None, None, None, None

# Initialize counter for unique ID generation
counter = 1

# Process each item, reshape data
reshaped_rows = []
for col in data.columns:
    manufacturer_name = col.split('_')[0]
    for item in data[col]:
        if pd.notna(item):
            brand, model, year, color = extract_info_improved(item)
            reshaped_rows.append({
                'Model_Manufacturer_name': manufacturer_name,
                'Collectible_item_full_name': item,
                'Brand': brand, 'Model': model, 'Year': year, 'Color': color,
                'id': f"{manufacturer_name}-{item}-{counter:04d}"
            })
            counter += 1

# Convert reshaped_rows into a DataFrame
df_b = pd.DataFrame(reshaped_rows)

# Reporting missing data and total entries
total_entries = len(df_b)
missing_brand = df_b['Brand'].isnull().sum()
missing_year = df_b['Year'].isnull().sum()
missing_color = df_b['Color'].isnull().sum()

print(f"Total number of entries: {total_entries}")
print(f"Number of entries with missing 'Brand': {missing_brand}")
print(f"Number of entries with missing 'Year': {missing_year}")
print(f"Number of entries with missing 'Color': {missing_color}")

# Display the DataFrame to verify results
display(df_b.head())


Total number of entries: 4045
Number of entries with missing 'Brand': 292
Number of entries with missing 'Year': 4045
Number of entries with missing 'Color': 3900


Model_Manufacturer_name         Collectible_item_full_name    Brand  \
0                  Altaya            Matra Simca Rancho 1978    Matra   
1                  Altaya  Citroen DS 19 Amsterdam Taxi 1958  Citroen   
2                  Altaya          Tuk tuk Bangkok taxi 1980      Tuk   
3                  Altaya                  Simca 1301 S 1972    Simca   
4                  Altaya   Peugeot 203 Casablanca Taxi 1960  Peugeot   

                     Model  Year Color  \
0        Simca Rancho 1978  None  None   
1  DS  Amsterdam Taxi 1958  None  None   
2    tuk Bangkok taxi 1980  None  None   
3              1301 S 1972  None  None   
4     Casablanca Taxi 1960  None  None   

                                              id  
0            Altaya-Matra Simca Rancho 1978-0001  
1  Altaya-Citroen DS 19 Amsterdam Taxi 1958-0002  
2          Altaya-Tuk tuk Bangkok taxi 1980-0003  
3                  Altaya-Simca 1301 S 1972-0004  
4   Altaya-Peugeot 203 Casablanca Taxi 1960-0005

## 2.5 Merging and Filling Gaps for Year and Color

This phase bridges the datasets that have been separately enhanced with year and color details. By merging these enriched tables, we systematically fill the informational gaps in our entries, ensuring a more comprehensive and complete dataset. This step is crucial for assembling a unified catalogue where each record is as detailed and accurate as possible, providing a seamless integration of temporal and visual data.

In [10]:
import pandas as pd

# Assuming df_a and df_b have been previously defined and loaded as shown in the updated Code A and Code B

# Merge df_a and df_b using the 'id' field
# We take all columns from df_a and only the 'Color' column from df_b
merged_data = pd.merge(df_a, df_b[['id', 'Color']], on='id', how='left')

# Reorder the columns to place 'id' before 'Model_Manufacturer_name'
column_order = ['id', 'Model_Manufacturer_name'] + [col for col in merged_data.columns if col not in ['id', 'Model_Manufacturer_name']]
merged_data = merged_data[column_order]

# We already have the Year information filled in df_a as needed, so we just need to add Color information
# Color from df_b will overwrite the Color in df_a where it exists
merged_data['Color'] = merged_data['Color'].combine_first(merged_data['Color'])

# Now you have a DataFrame with all information merged where the 'Year' comes from df_a and 'Color' from df_b
# Let's calculate missing data statistics to ensure everything aligns with expectations
total_entries = len(merged_data)
missing_brand = merged_data['Brand'].isnull().sum()
missing_year = merged_data['Year'].isnull().sum()
missing_color = merged_data['Color'].isnull().sum()

print(f"Total number of entries: {total_entries}")
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brand}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_year}")
print(f"Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': {missing_color}")

# Export the merged data to an Excel file
merged_data.to_excel('cleaned_data_cars_catalogue_v5_final.xlsx', index=False)

# Display the DataFrame to verify results
display(merged_data.head())


Total number of entries: 4045
Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 310
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 3550
Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': 3900


id Model_Manufacturer_name  \
0            Altaya-Matra Simca Rancho 1978-0001                  Altaya   
1  Altaya-Citroen DS 19 Amsterdam Taxi 1958-0002                  Altaya   
2          Altaya-Tuk tuk Bangkok taxi 1980-0003                  Altaya   
3                  Altaya-Simca 1301 S 1972-0004                  Altaya   
4   Altaya-Peugeot 203 Casablanca Taxi 1960-0005                  Altaya   

          Collectible_item_full_name    Brand                 Model  Year  \
0            Matra Simca Rancho 1978    Matra          Simca Rancho  1978   
1  Citroen DS 19 Amsterdam Taxi 1958  Citroen  DS 19 Amsterdam Taxi  1958   
2          Tuk tuk Bangkok taxi 1980      Tuk      tuk Bangkok taxi  1980   
3                  Simca 1301 S 1972    Simca                1301 S  1972   
4   Peugeot 203 Casablanca Taxi 1960  Peugeot   203 Casablanca Taxi  1960   

  Color  
0  None  
1  None  
2  None  
3  None  
4  None

In [11]:
import pandas as pd

# Assuming 'merged_data' is your DataFrame containing the data
# Extract unique brands and sort them alphabetically
unique_brands = merged_data['Brand'].dropna().unique()
unique_brands_sorted = sorted(unique_brands)

# Display the total number of unique brands
total_unique_brands = len(unique_brands_sorted)
print(f"Total number of unique brand names: {total_unique_brands}")

# Show only the first 15 and last 15 sorted unique brands
print("\nUnique Brands (First 15):")
for brand in unique_brands_sorted[:15]:
    print(brand)

print("\nUnique Brands (Last 15):")
for brand in unique_brands_sorted[-15:]:
    print(brand)


Total number of unique brand names: 466

Unique Brands (First 15):

"Commuter"
"Quartermaster"
ABT
AC
AEC
AGF
ALFA
AMC
AMD
AMG
ARO
ASTON
AWZ
Abarth

Unique Brands (Last 15):
ВНИИТЭ-ПТ
Волга
ГАЗ
ГАЗ-А
ЗИС
ЗиС
Иж
Москвич
НАТИ-АР
РАФ
СМЗ
СХТЗ-НАТИ
УАЗ
ХТЗ
“Monkees”


In [12]:
import pandas as pd

# Define the correction dictionary with wrong and correct spellings
corrections = {
    "ALFA": "Alfa",
    "ASTON": "Aston",
    "AWZ": "AWZ",
    "Autobianch": "Autobianchi",
    "BUGATTI": "Bugatti",
    "Betliet": "Berliet",
    "Brbham": "Brabham",
    "CHEVROLET": "Chevrolet",
    "CHRYSLER": "Chrysler",
    "CITROËN": "Citroën",
    "Cadillac": "Cadillac",
    "Citroen": "Citroën",
    "Duesemberg": "Duesenberg",
    "FERRARI": "Ferrari",
    "FIAT": "Fiat",
    "FORD": "Ford",
    "GAZ": "GAZ",
    "HUMMER": "Hummer",
    "ISO": "ISO",
    "ISUZU": "Isuzu",
    "Iveco": "IVECO",
    "LAMBORGHINI": "Lamborghini",
    "LOLA": "Lola",
    "MERCEDES-BENZ": "Mercedes-Benz",
    "Mercede-Benz": "Mercedes-Benz",
    "MINI": "Mini",
    "Maserati": "Maserati",
    "Mclaren": "McLaren",
    "Moskvitch": "Москвич",
    "Moskwitch": "Москвич",
    "Oldsmobil": "Oldsmobile",
    "PANHARD": "Panhard",
    "PORSCHE": "Porsche",
    "Plimouth": "Plymouth",
    "RENAULT": "Renault",
    "Red": "Red",
    "Saab": "SAAB",
    "SAVA": "Sava",
    "SAVIEM": "Saviem",
    "SEAT": "Seat",
    "SIAM": "Siam",
    "SIMCA": "Simca",
    "VW": "Volkswagen",
    "ЗИС": "ЗИС"
}

# Apply corrections to the 'Model_Manufacturer_name' column
merged_data['Model_Manufacturer_name'] = merged_data['Model_Manufacturer_name'].replace(corrections)

# Reorder the columns to place 'id' before 'Model_Manufacturer_name'
column_order = ['id', 'Model_Manufacturer_name'] + [col for col in merged_data.columns if col not in ['id', 'Model_Manufacturer_name']]
merged_data = merged_data[column_order]

# Now you have a DataFrame with corrected brand names
# Let's calculate missing data statistics to ensure everything aligns with expectations
total_entries = len(merged_data)
missing_brand = merged_data['Brand'].isnull().sum()
missing_year = merged_data['Year'].isnull().sum()
missing_color = merged_data['Color'].isnull().sum()

print(f"Total number of entries: {total_entries}")
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brand}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_year}")
print(f"Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': {missing_color}")

# Export the merged data to an Excel file
merged_data.to_excel('cleaned_data_cars_catalogue_v5.1_final.xlsx', index=False)

# Display the DataFrame to verify results
display(merged_data.head())  # Modify this to display(merged_data) if you want to see the entire DataFrame


Total number of entries: 4045
Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 310
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 3550
Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': 3900


id Model_Manufacturer_name  \
0            Altaya-Matra Simca Rancho 1978-0001                  Altaya   
1  Altaya-Citroen DS 19 Amsterdam Taxi 1958-0002                  Altaya   
2          Altaya-Tuk tuk Bangkok taxi 1980-0003                  Altaya   
3                  Altaya-Simca 1301 S 1972-0004                  Altaya   
4   Altaya-Peugeot 203 Casablanca Taxi 1960-0005                  Altaya   

          Collectible_item_full_name    Brand                 Model  Year  \
0            Matra Simca Rancho 1978    Matra          Simca Rancho  1978   
1  Citroen DS 19 Amsterdam Taxi 1958  Citroen  DS 19 Amsterdam Taxi  1958   
2          Tuk tuk Bangkok taxi 1980      Tuk      tuk Bangkok taxi  1980   
3                  Simca 1301 S 1972    Simca                1301 S  1972   
4   Peugeot 203 Casablanca Taxi 1960  Peugeot   203 Casablanca Taxi  1960   

  Color  
0  None  
1  None  
2  None  
3  None  
4  None

In [13]:
import pandas as pd
import uuid  # Importing the UUID module to generate unique IDs

# Assuming df_a and df_b have been previously defined and loaded as shown in the updated Code A and Code B

# Merge df_a and df_b using the 'id' field
# We take all columns from df_a and only the 'Color' column from df_b
merged_data = pd.merge(df_a, df_b[['id', 'Color']], on='id', how='left')

# Reorder the columns to place 'id' before 'Model_Manufacturer_name'
column_order = ['id', 'Model_Manufacturer_name'] + [col for col in merged_data.columns if col not in ['id', 'Model_Manufacturer_name']]
merged_data = merged_data[column_order]

# We already have the Year information filled in df_a as needed, so we just need to add Color information
# Color from df_b will overwrite the Color in df_a where it exists
merged_data['Color'] = merged_data['Color'].combine_first(merged_data['Color'])

# Now you have a DataFrame with all information merged where the 'Year' comes from df_a and 'Color' from df_b

# Generate unique IDs using UUID
total_entries = len(merged_data)
merged_data['Unique_ID'] = [str(uuid.uuid4()) for _ in range(total_entries)]

# Let's calculate missing data statistics to ensure everything aligns with expectations
missing_brand = merged_data['Brand'].isnull().sum()
missing_year = merged_data['Year'].isnull().sum()
missing_color = merged_data['Color'].isnull().sum()

print(f"Total number of entries: {total_entries}")
print(f"Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': {missing_brand}")
print(f"Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': {missing_year}")
print(f"Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': {missing_color}")

# Export the merged data to an Excel file
merged_data.to_excel('cleaned_data_cars_catalogue_v5_final_with_ids.xlsx', index=False)

# Display the DataFrame to verify results
display(merged_data.head())  # Modify this to display(merged_data) if you want to see the entire DataFrame


Total number of entries: 4045
Number of entries with missing 'Brand', excluding missing 'Model_Manufacturer_name': 310
Number of entries with missing 'Year', excluding missing 'Model_Manufacturer_name': 3550
Number of entries with missing 'Color', excluding missing 'Model_Manufacturer_name': 3900


id Model_Manufacturer_name  \
0            Altaya-Matra Simca Rancho 1978-0001                  Altaya   
1  Altaya-Citroen DS 19 Amsterdam Taxi 1958-0002                  Altaya   
2          Altaya-Tuk tuk Bangkok taxi 1980-0003                  Altaya   
3                  Altaya-Simca 1301 S 1972-0004                  Altaya   
4   Altaya-Peugeot 203 Casablanca Taxi 1960-0005                  Altaya   

          Collectible_item_full_name    Brand                 Model  Year  \
0            Matra Simca Rancho 1978    Matra          Simca Rancho  1978   
1  Citroen DS 19 Amsterdam Taxi 1958  Citroen  DS 19 Amsterdam Taxi  1958   
2          Tuk tuk Bangkok taxi 1980      Tuk      tuk Bangkok taxi  1980   
3                  Simca 1301 S 1972    Simca                1301 S  1972   
4   Peugeot 203 Casablanca Taxi 1960  Peugeot   203 Casablanca Taxi  1960   

  Color                             Unique_ID  
0  None  179e9b00-9842-4e01-a31f-8ee1179a18df  
1  None  761cf7d7-a87c-4c9a-9a91-b07cde4bbde2  
2  None  de14e570-ac1f-40f5-942d-3123f96c1a3a  
3  None  d269d5a6-26dd-46ad-a6f5-aa1f537d6ad1  
4  None  2b977385-ffc1-434f-a0fc-8be658c4e719